### PR TITLE
Make flakeheaven work with flake8 4.0.x

### PR DESCRIPTION
**Pull request by mcarans**
Updated for code changes in flake8 4.0.x
Also updated for some plugin changes that broke tests
Tests run using pytest locally all pass
Documentation updated (for now I link directly to the .md files on GitHub)
Also tested in a real project with pyproject.toml:

```toml
[tool.flakeheaven]
# make output nice
format = "grouped"
max_line_length = 79
# show line of source code in output
show_source = true

    # list of plugins and rules for them
    [tool.flakeheaven.plugins]
    mccabe = ["+*"]
    pycodestyle = ["+*", "-E203", "-E501", "-W503"]
    pyflakes = ["+*"]
    pylint = ["+*"]
    flake8-isort = ["+*"]
    flake8-black = ["+*"]
...
[testenv:lint]
wheel_build_env = py310
deps =
    git+git://github.com/mcarans/flakeheaven.git#egg=flakeheaven
    flake8
    flake8-isort
    flake8-black
commands =
    flakeheaven lint src tests

```

### CODE REVIEW

**Comment by mcarans** [comment]
@pwoolvett @lyz-code @jarojasm95 I don't seem to have an option to add you as reviewers. I'm not sure why. Please can you take a look at the PR.

File: README.md [hunk 1]
```diff
@@ -1,56 +1,52 @@
...
 + [PyLint](https://github.com/PyCQA/pylint) integration.
-+ [Powerful GitLab support](https://flakehell.readthedocs.io/formatters.html#gitlab).
++ [Powerful GitLab support](https://github.com/mcarans/flakeheaven/blob/master/docs/formatters.md#gitlab).
 + Codes management:
     + Manage codes per plugin.
     + Enable and disable plugins and codes by wildcard.
-    + [Show codes for installed plugins](https://flakehell.readthedocs.io/commands/plugins.html).
-    + [Show all messages and codes for a plugin](https://flakehell.readthedocs.io/commands/codes.html).
+    + [Show codes for installed plugins](https://github.com/mcarans/flakeheaven/blob/master/docs/commands/plugins.md).
+    + [Show all messages and codes for a plugin](https://github.com/mcarans/flakeheaven/blob/master/docs/commands/codes.md).
```
Comment (jarojasm95) [inline README.md:27]: all the above links should probably point to `https://github.com/flakeheaven/flakeheaven` I think instead of `https://github.com/mcarans/flakeheaven`

File: README.md [hunk 2]
```diff
@@ -1,56 +1,52 @@
...
 ## Usage
 
 First of all, let's create `pyproject.toml` config:
 
 ```toml
-[tool.flakehell]
+[tool.flakeheaven]
 # optionally inherit from remote config (or local if you want)
-base = "https://raw.githubusercontent.com/life4/flakehell/master/pyproject.toml"
+base = "https://raw.githubusercontent.com/life4/flakeheaven/master/pyproject.toml"
```
Comment (jarojasm95) [inline README.md:49]:
```suggestion
base = "https://raw.githubusercontent.com/flakeheaven/flakeheaven/master/pyproject.toml"
```

File: README.md [hunk 3]
```diff
@@ -75,34 +71,34 @@ flake8-docstrings = ["-*"]
...
 
 ```bash
-flakehell lint
+flakeheaven lint
 ```
 
 This command accepts all the same arguments as Flake8.
 
-Read [flakehell.readthedocs.io](https://flakehell.readthedocs.io/) for more information.
+Read [github.com/mcarans/flakeheaven/blob/master/docs](https://github.com/mcarans/flakeheaven/blob/master/docs/) for more information.
```
Comment (jarojasm95) [inline README.md:101]:
```suggestion
Read [github.com/flakeheaven/flakeheaven/blob/master/docs](https://github.com/flakeheaven/flakeheaven/blob/master/docs/) for more information.
```

File: docs/conf.py [hunk 4]
```diff
@@ -53,18 +53,18 @@
     # 'color': '#2c3e50',
     'github_banner': 'true',
     'github_user': 'life4',
-    'github_repo': 'flakehell',
+    'github_repo': 'flakeheaven',
     'github_type': 'star',
 
     'extra_nav_links': {
-        'GitHub repository': 'https://github.com/life4/flakehell',
+        'GitHub repository': 'https://github.com/life4/flakeheaven',
```
Comment (jarojasm95) [inline docs/conf.py:60]:
```suggestion
        'GitHub repository': 'https://github.com/flakeheaven/flakeheaven',
```

File: docs/config.md [hunk 5]
```diff
@@ -45,15 +45,15 @@ path can be either a path prefix (from the project root) or a [glob pattern](htt
 Option `base` allows to specify base config from which you want to inherit this one. It can be path to local config or remote URL. You can specify one path or list of paths as well. For example:
 
 ```toml
-base = ["https://raw.githubusercontent.com/life4/flakehell/master/pyproject.toml", ".flakehell.toml"]
+base = ["https://raw.githubusercontent.com/life4/flakeheaven/master/pyproject.toml", ".flakeheaven.toml"]
```
Comment (jarojasm95) [inline docs/config.md:48]:
```suggestion
base = ["https://raw.githubusercontent.com/flakeheaven/flakeheaven/master/pyproject.toml", ".flakeheaven.toml"]
```

File: docs/config.md [hunk 6]
```diff
@@ -84,9 +84,9 @@ FlakeHell doesn't support some flake8 option by design. Flake8 has a long histor
 ## Example
 
 ```toml
-[tool.flakehell]
+[tool.flakeheaven]
 # optionally inherit from remote config (or local if you want)
-base = "https://raw.githubusercontent.com/life4/flakehell/master/pyproject.toml"
+base = "https://raw.githubusercontent.com/life4/flakeheaven/master/pyproject.toml"
```
Comment (jarojasm95) [inline docs/config.md:89]:
```suggestion
base = "https://raw.githubusercontent.com/flakeheaven/flakeheaven/master/pyproject.toml"
```

File: docs/index.md [hunk 7]
```diff
@@ -1,30 +1,30 @@
...
 + [PyLint](https://github.com/PyCQA/pylint) integration.
-+ [Powerful GitLab support](https://flakehell.readthedocs.io/formatters.html#gitlab).
++ [Powerful GitLab support](https://github.com/mcarans/flakeheaven/blob/master/docs/formatters.md#gitlab).
 + Codes management:
     + Manage codes per plugin.
     + Enable and disable plugins and codes by wildcard.
-    + [Show codes for installed plugins](https://flakehell.readthedocs.io/commands/plugins.html).
-    + [Show all messages and codes for a plugin](https://flakehell.readthedocs.io/commands/codes.html).
+    + [Show codes for installed plugins](https://github.com/mcarans/flakeheaven/blob/master/docs/commands/plugins.md).
+    + [Show all messages and codes for a plugin](https://github.com/mcarans/flakeheaven/blob/master/docs/commands/codes.md).
```
Comment (jarojasm95) [inline docs/index.md:20]: links should probs be changed from `mcarans` -> `flakeheaven`

File: docs/troubleshooting.md [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
+FlakeHeaven has flake8 deep in its core but they are different. So, when migrating from flake8 there are some details that can be unexpected from your previous experience. So, let's talk how to turn such details from your enemies to your best friends.
 
 ## Flake8 configurations are ignored
 
-FlakeHell reads flake8 configs and supports most of the options. However, some options (like `ignore` and `select`) ignored by design. The motivation is to get rid of over-complicated and unstable way how flake8 select and deselect plugins and provide a better way to do so. See [ignored options](https://flakehell.readthedocs.io/config.html#ignored-options) for details.
+FlakeHeaven reads flake8 configs and supports most of the options. However, some options (like `ignore` and `select`) ignored by design. The motivation is to get rid of over-complicated and unstable way how flake8 select and deselect plugins and provide a better way to do so. See [ignored options](https://github.com/mcarans/flakeheaven/blob/master/docs/config.md#ignored-options) for details.
```
Comment (jarojasm95) [inline docs/troubleshooting.md:7]:
```suggestion
FlakeHeaven reads flake8 configs and supports most of the options. However, some options (like `ignore` and `select`) ignored by design. The motivation is to get rid of over-complicated and unstable way how flake8 select and deselect plugins and provide a better way to do so. See [ignored options](https://github.com/flakeheaven/flakeheaven/blob/master/docs/config.md#ignored-options) for details.
```

File: flakeheaven/_logic/_extractors.py [hunk 9]
```diff
@@ -220,10 +221,13 @@ def extract_flake8_pytest_style() -> Dict[str, str]:
 
 
 def extract_flake8_annotations_complexity() -> Dict[str, str]:
-    from flake8_annotations_complexity.checker import AnnotationsComplexityChecker
```
Comment (jarojasm95) [inline flakeheaven/_logic/_extractors.py:223]: is this class/module not available anymore? or what is the reason for this change?

File: flakeheaven/_logic/_extractors.py [hunk 10]
```diff
@@ -401,3 +405,23 @@ def extract_wemake_python_styleguide() -> Dict[str, str]:
             code = 'WPS' + str(checker.code).zfill(3)
             codes[code] = checker.error_template
     return codes
+
+
+def extract_flake8_pie() -> Dict[str, str]:
```
Comment (jarojasm95) [inline flakeheaven/_logic/_extractors.py:410]: new feature? 🥳

File: flakeheaven/_logic/_snapshot.py [hunk 11]
```diff
@@ -11,7 +11,7 @@
 from flake8.options.manager import OptionManager
 
 
-CACHE_PATH = Path(os.environ.get('FLAKEHELL_CACHE', Path.home() / '.cache' / 'flakehell'))
+CACHE_PATH = Path(os.environ.get('FLAKEHELL_CACHE', Path.home() / '.cache' / 'flakeheaven'))
```
Comment (jarojasm95) [inline flakeheaven/_logic/_snapshot.py:14]:
```suggestion
CACHE_PATH = Path(os.environ.get('FLAKEHEAVEN_CACHE', Path.home() / '.cache' / 'flakeheaven'))
```

File: pyproject.toml [hunk 12]
```diff
@@ -90,22 +90,22 @@ classifiers=[
 ]
 
 [tool.flit.metadata.urls]
-Documentation = "https://flakehell.readthedocs.io/"
+Documentation = "https://github.com/mcarans/flakeheaven/blob/master/docs/"
```
Comment (jarojasm95) [inline pyproject.toml:93]:
```suggestion
Documentation = "https://github.com/flakeheaven/flakeheaven/blob/master/docs/"
```

**Review by jarojasm95 — COMMENTED** [review]
I think a couple links still need to be updated - but this is great, thanks!

File: README.md [hunk 13]
```diff
@@ -1,56 +1,52 @@
...
+[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
 
-# ![FlakeHell](./assets/logo.png)
+FlakeHeaven is a [Flake8](https://gitlab.com/pycqa/flake8) wrapper to make it cool.
 
-[![PyPI version](https://badge.fury.io/py/flakehell.svg)](https://badge.fury.io/py/flakehell)
-[![Build Status](https://cloud.drone.io/api/badges/life4/flakehell/status.svg)](https://cloud.drone.io/life4/flakehell)
-[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
-[![Documentation](https://readthedocs.org/projects/flakehell/badge/?version=latest)](https://flakehell.readthedocs.io/)
+This project is a fork of [FlakeHell](https://github.com/life4/flakehell). FlakeHell and the official fork of it at
```
Comment (lyz-code) [inline README.md:7]: I'd add some note that this is now the new official fork maintained by the community, maybe pointing to the issue of flakehell where we decided to give the step forward

File: flakeheaven/_logic/_extractors.py [hunk 14]
```diff
@@ -401,3 +405,23 @@ def extract_wemake_python_styleguide() -> Dict[str, str]:
...
+    import flake8_pie
+    from inspect import getsource
```
Comment (lyz-code) [inline flakeheaven/_logic/_extractors.py:412]: I'd move the imports to the top of the file

**Review by lyz-code — CHANGES_REQUESTED** [review]
Thanks for all the work :)

File: docs/conf.py [hunk 15]
```diff
@@ -53,18 +53,18 @@
     # 'color': '#2c3e50',
     'github_banner': 'true',
     'github_user': 'life4',
```
Comment (pwoolvett) [inline docs/conf.py:55]: 'github_user': 'flakeheaven',

File: README.md (hunk 13 above)
Comment (pwoolvett) [reply README.md:7]: what do you mean by official? I mean, it's as official as any other fork!

Comment (lyz-code) [reply README.md:7]: Maybe official is not the correct wording, what I meant is that this is the fork that is going to be maintained by the community of all the existent forks, and that the previous maintainer agreed to do this migration.

I'd find it interesting from a user stand point, as right now there are a lot of forks and tools with similar names that do the same. It will help them choose

File: flakeheaven/_logic/_extractors.py (hunk 14 above)
Comment (jarojasm95) [reply flakeheaven/_logic/_extractors.py:412]: I could be wrong, but I think this was intentional so that the module is not dependent on this `flake8_pie` package unless this function is called, the same pattern seems repeated across this file

File: flakeheaven/_logic/_extractors.py (hunk 9 above)
Comment (mcarans) [reply flakeheaven/_logic/_extractors.py:223]: The plugin code had changed so the existing test was failing. The class exists:
https://github.com/best-doctor/flake8-annotations-complexity/blob/master/flake8_annotations_complexity/checker.py

However AnnotationsComplexityChecker._error_message_template no longer exists.

File: README.md [hunk 16]
```diff
@@ -75,34 +71,34 @@ flake8-docstrings = ["-*"]
...
 
 ```bash
-flakehell lint
+flakeheaven lint
 ```
 
 This command accepts all the same arguments as Flake8.
 
-Read [flakehell.readthedocs.io](https://flakehell.readthedocs.io/) for more information.
+Read [github.com/mcarans/flakeheaven/blob/master/docs](https://github.com/flakeheaven/flakeheaven/blob/master/docs/) for more information.
```
Comment (jarojasm95) [inline README.md:101]:
```suggestion
Read [github.com/flakeheaven/flakeheaven/blob/master/docs](https://github.com/flakeheaven/flakeheaven/blob/master/docs/) for more information.
```

File: flakeheaven/_logic/_extractors.py (hunk 14 above)
Comment (mcarans) [reply flakeheaven/_logic/_extractors.py:412]: I will leave this unchanged because that is how the other plugins were handled and @jarojasm95 is likely correct about it being intentional.

**Comment by mcarans** [comment]
Thanks so much for all the helpful suggestions. I have updated accordingly.

**Review by jarojasm95 — COMMENTED** [review]
looks great! I think one more small change is needed and then it's g2g

File: README.md (hunk 16 above)
Comment (mcarans) [reply README.md]: Thanks for catching that one!

**Review by jarojasm95 — APPROVED** [review]
lgtm - thanks for putting this together!

**Comment by pwoolvett** [comment]
Excelent!!!

I haven't looked into your actual code yet (!), just ran a simple pass on a file and it seems fine. However,

I ran the tests from 4f0f98f334eaeadfaba90b9f3dcbf8e88f050e34
* [ ] docs wont build: `dropqa.md:document isn't included in any toctree` (the `-W` flag)
* [ ] mypy: `Found 6 errors in 3 files (checked 45 source files)` (had to include `mypy` in the dev group)
* [ ] isort fails (probably old syntax, using the `-y` flag). We could either update the command or pin to a former version
*  [ ] `flake8-strict` is broken with recent `black` - we require `black==21.10b0`
* [ ] `flakeheaven plugins` fails (pylint plugin reports no version)
* [ ] `python 3.6` fails (no `__annotations__` in `future`, used by `flake8_functions` -> `mr_proper` NOTE: I'm ok with dropping python 3.6, but it seems fixable with a version pinning to `0.0.4`. However that conficts with the black pinning above for `mypy-extensions`...

1. Is any of the above related to this PR? otherwise, we could fix those in another PR
2. What do you think about adding MWE github ci se we can pickup from this startingpoint for the next contributions (see below diff and file)
3. If all is fine, I'd merge to master, tag 1.0.0 and send to pypi


```diff
diff --git a/pyproject.toml b/pyproject.toml
index c3f9e30..2c788cc 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -137,7 +137,7 @@ dev = [
     "flake8-executable",
     "flake8-expression-complexity",
     "flake8-fixme",
-    "flake8-functions",
+    "flake8-functions",  #  "mr_proper==0.0.4",  # support py36
     "flake8-future-import",
     "flake8-import-order",
     "flake8-isort",
@@ -158,7 +158,7 @@ dev = [
     "flake8-scrapy",
     "flake8-spellcheck",
     "flake8-sql",
-    "flake8-strict",
+    "flake8-strict",  "black==21.10b0", # https://github.com/smarkets/flake8-strict/issues/41
     "flake8-string-format",
     "flake8-tidy-imports",
     "flake8-todo",
@@ -172,6 +172,7 @@ dev = [
     "wemake-python-styleguide",
 
     "pytest",
+    "mypy",
     "isort[pyproject]",
 ]
```
```yaml
#.github/workflows/ci.yaml
name: CI
on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master

jobs:
  default:
    name: "${{ matrix.step.env }} '${{ matrix.step.command }}' @ py${{ matrix.step.python }})"
    runs-on: ubuntu-latest
    strategy:
      matrix:
        step: [
          # original from drone CI startlark file
          # {env: "pytest", python: "3.6", extras: "main,dev", command: "pytest tests/"},  # disable or drop support?
          {env: "pytest", python: "3.7", extras: "main,dev", command: "pytest tests/"},
          {env: "pytest", python: "3.8", extras: "main,dev", command: "pytest tests/"},
          {env: "pytest", python: "3.9", extras: "main,dev", command: "pytest tests/"},
          {env: "main", python: "3.8", extras: "main,dev", command: "flakeheaven lint"},
          {env: "typing", python: "3.8", extras: "main,dev", command: "mypy --ignore-missing-imports --allow-redefinition flakeheaven/"},
          # additional
          {env: "docs", python: "3.8", extras: "docs", command: "sphinx-build -W docs docs/build"},
          {env: "isort", python: "3.8", extras: "main,dev", command: "isort -rc -y"},
        ]
    container:
      image: python:${{ matrix.step.python }}-alpine
      env:
        COVERAGE_FILE: ".coverage.${{ matrix.step.env }}.${{ matrix.step.python }}"
    steps:
      - name: Check out repository
        uses: actions/checkout@v2
      - name: Setup env
        run: |
          apk add \
            gcc \
            libc-dev
      - name: Install with deps
        run: |
          pip install .[${{ matrix.step.extras }}]

      - name: Run command
        run: |
          ${{ matrix.step.command }}

```

**Comment by mcarans** [comment]
@pwoolvett The mypy and isort one might be related. I can check that. Where did you see this failure "flakeheaven plugins fails (pylint plugin reports no version)"? The others are unrelated I think. How did you run the tests? All I have been doing is running pytest from PyCharm where all the tests in the tests folder (188 of them) pass with Python 3.10.

**Comment by mcarans** [comment]
@pwoolvett I added the CI and pyproject.toml changes you suggested. I'm happy for you to go ahead with merge and release whenever you're ready.

**Comment by pwoolvett** [comment]
> (The _config.py ones went away after I ran `mypy --install-types` leaving 2 errors)

So it would be enough to run that in CI before checking types?

> Where did you see this failure "flakeheaven plugins fails (pylint plugin reports no version)"?

* run the command `flakeheaven plugins`. the "pylint plugin reports no version" is what I think is going on from the error:
  ```py
    File "lib/python3.8/site-packages/flakeheaven/commands/_plugins.py", line 17, in <genexpr>
    version_width = max(8, max(len(p['version']) for p in plugins))
  ```
  (no idea if its related to the flake4 update...)

> The others are unrelated I think. How did you run the tests? All I have been doing is running pytest from PyCharm where all the tests in the tests folder (188 of them) pass with Python 3.10.

* I'm using [act](https://github.com/nektos/act) to run the CI workflow (along with the `pyproject.toml` diff, now included in your latest commit)

Make flakeheaven great again!

**Comment by pwoolvett** [comment]
In the meantime, Its on [testpypi](https://test.pypi.org/project/flakeheaven/0.11.0/)

**Comment by pwoolvett** [comment]
> @pwoolvett I added the CI and pyproject.toml changes you suggested. I'm happy for you to go ahead with merge and release whenever you're ready.

@mcarans please set `__version__` to 0.11.0 in `_version` and we're good (to avoid conflicts with previous releases)

**Comment by mcarans** [comment]
@pwoolvett I've updated it
